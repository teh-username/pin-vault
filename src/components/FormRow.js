import React from 'react';
import { FormLabel, FormInput } from 'react-native-elements';

const FormRow = ({ label, name, value, onInputChange, editable = true }) => [
  <FormLabel key={`${name}_label`}>{label}</FormLabel>,
  <FormInput
    key={`${name}input`}
    value={value}
    onChangeText={data => onInputChange(name, data)}
    placeholder={`Please enter ${label.toLowerCase()}...`}
    editable={editable}
  />,
];

export default FormRow;
