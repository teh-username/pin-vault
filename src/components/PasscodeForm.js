import React from 'react';
import { View, StyleSheet } from 'react-native';
import { Button, FormValidationMessage } from 'react-native-elements';

import FormRow from './FormRow';

class PasscodeForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      passcode: '',
      repeatPasscode: '',
      error: null,
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  handleFormSubmit() {
    const { passcode, repeatPasscode } = this.state;

    if (passcode !== repeatPasscode) {
      this.setState({
        error: "Passcodes doesn't match",
      });
      return;
    }

    this.props.handlePasscodeSubmit(this.state.passcode);
    this.setState({
      error: null,
    });
  }

  handleInputChange(stateKey, data) {
    this.setState({
      [stateKey]: data,
    });
  }

  render() {
    const { passcode, repeatPasscode, error } = this.state;
    return (
      <View>
        <FormRow
          label="Passcode"
          name="passcode"
          value={passcode}
          onInputChange={this.handleInputChange}
        />
        <FormRow
          label="Confirm Passcode"
          name="repeatPasscode"
          value={repeatPasscode}
          onInputChange={this.handleInputChange}
        />
        {error ? <FormValidationMessage>{error}</FormValidationMessage> : null}
        <Button
          title="Set Passcode"
          icon={{ name: 'add' }}
          onPress={this.handleFormSubmit}
          buttonStyle={styles.buttonStyle}
          disabled={!passcode || !repeatPasscode}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  buttonStyle: {
    top: 40,
    backgroundColor: '#669AE0',
  },
});

export default PasscodeForm;
