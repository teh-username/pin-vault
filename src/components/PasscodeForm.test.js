import React from 'react';
import { FormValidationMessage } from 'react-native-elements';

import PasscodeForm from './PasscodeForm';
import FormRow from './FormRow';

describe('PasscodeForm component test', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<PasscodeForm />);
    expect(wrapper.find(FormRow).length).toEqual(2);
    expect(
      wrapper
        .find(FormRow)
        .at(0)
        .prop('label')
    ).toEqual('New Passcode');
    expect(wrapper).toMatchSnapshot();
  });

  it('should render old passcode input if user has previously set passcode', () => {
    const wrapper = shallow(<PasscodeForm oldPasscode="passcode" />);
    expect(wrapper.find(FormRow).length).toEqual(3);
    expect(
      wrapper
        .find(FormRow)
        .at(0)
        .prop('label')
    ).toEqual('Current Passcode');
    expect(wrapper).toMatchSnapshot();
  });

  it('should render error field if there are any errors', () => {
    const wrapper = shallow(<PasscodeForm />);
    wrapper.setState({ error: 'error' });
    expect(wrapper.find(FormValidationMessage).prop('children')).toEqual(
      'error'
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should return an error message if passcode and repeat passcode do not match', () => {
    const wrapper = shallow(<PasscodeForm />);
    wrapper.setState({
      passcode: 'yes',
      repeatPasscode: 'no',
    });
    wrapper.instance().handleFormSubmit();
    expect(wrapper.state('error')).toEqual("Passcodes don't match");
  });

  it('should return an error message if verification passcode does not match the current passcode', () => {
    const wrapper = shallow(
      <PasscodeForm oldPasscode="5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8" />
    );
    wrapper.setState({
      passcode: 'yes',
      repeatPasscode: 'yes',
      currentPasscode: 'notPassword',
    });
    wrapper.instance().handleFormSubmit();
    expect(wrapper.state('error')).toEqual('Invalid current passcode');
  });

  it('should call callback if user has matching passcode and has not previously set a passcode', () => {
    const mockCallback = jest.fn();
    const wrapper = shallow(
      <PasscodeForm handlePasscodeSubmit={mockCallback} />
    );
    wrapper.setState({
      passcode: 'yes',
      repeatPasscode: 'yes',
    });

    wrapper.instance().handleFormSubmit();
    expect(mockCallback.mock.calls.length).toBe(1);
    expect(mockCallback.mock.calls[0][0]).toBe('yes');
  });

  it('should call callback if user has matching passcode and has a correct old passcode ', () => {
    const mockCallback = jest.fn();
    const wrapper = shallow(
      <PasscodeForm
        handlePasscodeSubmit={mockCallback}
        oldPasscode="5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8"
      />
    );
    wrapper.setState({
      passcode: 'nope',
      repeatPasscode: 'nope',
      currentPasscode: 'password',
    });
    wrapper.instance().handleFormSubmit();
    expect(mockCallback.mock.calls.length).toBe(1);
    expect(mockCallback.mock.calls[0][0]).toBe('nope');
  });
});
