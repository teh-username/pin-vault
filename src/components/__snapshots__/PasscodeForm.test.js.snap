// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasscodeForm component test should render as expected 1`] = `
<View>
  <FormRow
    label="New Passcode"
    name="passcode"
    onInputChange={[Function]}
    value=""
  />
  <FormRow
    label="Confirm New Passcode"
    name="repeatPasscode"
    onInputChange={[Function]}
    value=""
  />
  <Button
    buttonStyle={
      Object {
        "backgroundColor": "#669AE0",
        "top": 40,
      }
    }
    disabled={true}
    icon={
      Object {
        "name": "add",
      }
    }
    onPress={[Function]}
    title="Set Passcode"
  />
</View>
`;

exports[`PasscodeForm component test should render error field if there are any errors 1`] = `
<View>
  <FormRow
    label="New Passcode"
    name="passcode"
    onInputChange={[Function]}
    value=""
  />
  <FormRow
    label="Confirm New Passcode"
    name="repeatPasscode"
    onInputChange={[Function]}
    value=""
  />
  <FormValidationMessage>
    error
  </FormValidationMessage>
  <Button
    buttonStyle={
      Object {
        "backgroundColor": "#669AE0",
        "top": 40,
      }
    }
    disabled={true}
    icon={
      Object {
        "name": "add",
      }
    }
    onPress={[Function]}
    title="Set Passcode"
  />
</View>
`;

exports[`PasscodeForm component test should render old passcode input if user has previously set passcode 1`] = `
<View>
  <FormRow
    label="Current Passcode"
    name="currentPasscode"
    onInputChange={[Function]}
    value=""
  />
  <FormRow
    label="New Passcode"
    name="passcode"
    onInputChange={[Function]}
    value=""
  />
  <FormRow
    label="Confirm New Passcode"
    name="repeatPasscode"
    onInputChange={[Function]}
    value=""
  />
  <Button
    buttonStyle={
      Object {
        "backgroundColor": "#669AE0",
        "top": 40,
      }
    }
    disabled={true}
    icon={
      Object {
        "name": "add",
      }
    }
    onPress={[Function]}
    title="Set Passcode"
  />
</View>
`;
